bindings/python/src/capability.rs

```diff
@@ -0,0 +1,90 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use pyo3::exceptions::PyAttributeError;
+use pyo3::prelude::*;
+use pyo3::types::PyString;
+
+#[pyclass(module = "opendal")]
+pub struct Capability(opendal::Capability);
```

Comment:
How about using `#[derive(FromPyObject)]` and `#[pyo3(transparent)]`?

Reply:
I think maybe the `#[derive(FromPyObject)]` and `#[pyo3(transparent)]` fous on the data flow from PythonObject to Rust Object? I'm not sure it would be works well when the data flow reversed.

In another hand, I would like to keep the same style with  other class like `Reader`/`Operator`

Reply:
> I would like to keep the same style with other class

I believe it's unnecessary. Unlike `Reader` and `Operator`, `Capability` doesn't store any pointers, just a collection of values.

Reply:
cc @messense WDYT?

Reply:
Another concern is that as `Capability` continues to expand, manually maintaining the mapping becomes increasingly challenging.

Reply:
> manually maintaining the mapping becomes increasingly challenging.
We need to maintain it manually(case we still got one step to add type hint)

Reply:
> How about using `#[derive(FromPyObject)]` and `#[pyo3(transparent)]`?

It's not applicable here because it's for extracting data from Python then convert to Rust struct/enum, what you want is really a derive macro for `IntoPy<T>` but that doesn't exist.

Reply:
So I think this PR should be ready to merge?

Reply:
No, I don't want to merge this PR as-is. Can we use `#[pyclass]` and `#[pyo3(get)]` or `get_all`? Or is there a way to build a `dataclasses` in Python?

Reply:
> Can we use `#[pyclass]` and `#[pyo3(get)]` or `get_all`?

You can do that if you make `opendal` core conditional depend on `pyo3` so you can apply `#[pyclass]` on `opendal::Capability` but IMO it's not worth the added complexity.

`#[pyclass]` macro can't not inspect the fields of `opendal::Capability` from opendal-python because `syn` only operates on token streams, it's a fundamental limitation of Rust derive macro.

Reply:
IMO we could add a test case to ensure that `Capability` in python is in sync with `Capability` in Rust. For example, first we serialize `opendal::Capability` to JSON to get all field names, then call `operator.capability().<field name>` in Python for every field.

Reply:
@messense PTAL